// clawless/dist/messaging/slackClient.js
import { App } from '@slack/bolt';
import { slackifyMarkdown } from 'slackify-markdown';
import { logInfo, logError } from '../utils/error.js';
function toSlackMrkdwn(text) {
    const normalized = String(text || '').replace(/\r\n/g, '\n');
    if (!normalized) {
        return normalized;
    }
    try {
        return slackifyMarkdown(normalized);
    }
    catch {
        return normalized;
    }
}
function splitTextIntoChunks(text, maxMessageLength) {
    const normalizedText = String(text || '');
    if (!normalizedText) {
        return [''];
    }
    if (normalizedText.length <= maxMessageLength) {
        return [normalizedText];
    }
    const chunks = [];
    let start = 0;
    while (start < normalizedText.length) {
        const end = Math.min(start + maxMessageLength, normalizedText.length);
        chunks.push(normalizedText.slice(start, end));
        start = end;
    }
    return chunks;
}
export class SlackMessageContext {
    event;
    app;
    typingIntervalMs;
    maxMessageLength;
    text;
    chatId;
    userId;
    typingInterval = null;
    liveMessageTextByTs = new Map();
    constructor(event, app, typingIntervalMs, maxMessageLength) {
        this.event = event;
        this.app = app;
        this.typingIntervalMs = typingIntervalMs;
        this.maxMessageLength = maxMessageLength;
        this.text = event.text || '';
        this.chatId = event.channel;
        this.userId = event.user;
    }
    startTyping() {
        // Slack doesn't have a direct typing indicator API
        // We'll use a no-op implementation to maintain interface compatibility
        const stopTyping = () => {
            if (this.typingInterval) {
                clearInterval(this.typingInterval);
                this.typingInterval = null;
            }
        };
        return stopTyping;
    }
    async sendText(text) {
        const formattedText = toSlackMrkdwn(text);
        const chunks = splitTextIntoChunks(formattedText, this.maxMessageLength);
        for (const chunk of chunks) {
            await this.app.client.chat.postMessage({
                channel: this.event.channel,
                text: chunk,
            });
        }
    }
    async startLiveMessage(initialText = '…') {
        const formattedText = toSlackMrkdwn(initialText || '…');
        const result = await this.app.client.chat.postMessage({
            channel: this.event.channel,
            text: formattedText,
        });
        const messageTs = result.ts;
        if (messageTs) {
            this.liveMessageTextByTs.set(messageTs, formattedText);
        }
        return messageTs;
    }
    async updateLiveMessage(messageTs, text) {
        try {
            const formattedText = toSlackMrkdwn(text || '…');
            const previousText = this.liveMessageTextByTs.get(messageTs);
            if (previousText === formattedText) {
                return;
            }
            await this.app.client.chat.update({
                channel: this.event.channel,
                ts: messageTs,
                text: formattedText,
            });
            this.liveMessageTextByTs.set(messageTs, formattedText);
        }
        catch (error) {
            // Slack may throw error if message hasn't changed
            // We'll ignore those errors
            if (!error.message?.includes('message_not_found') && !error.message?.includes('cant_update_message')) {
                throw error;
            }
        }
    }
    async finalizeLiveMessage(messageTs, text) {
        const finalText = toSlackMrkdwn(text || 'No response received.');
        const chunks = splitTextIntoChunks(finalText, this.maxMessageLength);
        const firstChunk = chunks[0] || 'No response received.';
        const previousText = this.liveMessageTextByTs.get(messageTs);
        try {
            if (previousText !== firstChunk) {
                await this.app.client.chat.update({
                    channel: this.event.channel,
                    ts: messageTs,
                    text: firstChunk,
                });
                this.liveMessageTextByTs.set(messageTs, firstChunk);
            }
        }
        catch (error) {
            const errorMessage = String(error?.message || '').toLowerCase();
            if (!errorMessage.includes('message_not_found') && !errorMessage.includes('cant_update_message')) {
                throw error;
            }
        }
        // Send remaining chunks as separate messages
        for (let index = 1; index < chunks.length; index += 1) {
            await this.app.client.chat.postMessage({
                channel: this.event.channel,
                text: chunks[index],
            });
        }
    }
    async removeMessage(messageTs) {
        try {
            await this.app.client.chat.delete({
                channel: this.event.channel,
                ts: messageTs,
            });
        }
        catch (error) {
            logError('Failed to delete Slack message:', error);
        }
    }
}
export class SlackMessagingClient {
    app;
    typingIntervalMs;
    maxMessageLength;
    messageHandlers = [];
    errorHandlers = [];
    constructor({ token, signingSecret, appToken, typingIntervalMs, maxMessageLength, }) {
        // Use Socket Mode if appToken is provided, otherwise use HTTP mode
        this.app = new App({
            token,
            signingSecret,
            ...(appToken
                ? {
                    socketMode: true,
                    appToken,
                }
                : {}),
        });
        this.typingIntervalMs = typingIntervalMs;
        this.maxMessageLength = maxMessageLength;
        // Listen to all message events
        this.app.message(async ({ message }) => {
            // Only handle regular messages (not bot messages)
            if (message.subtype === undefined && 'text' in message) {
                const slackEvent = message;
                const messageContext = new SlackMessageContext(slackEvent, this.app, this.typingIntervalMs, this.maxMessageLength);
                for (const handler of this.messageHandlers) {
                    try {
                        await Promise.resolve(handler(messageContext));
                    }
                    catch (error) {
                        logError('Slack message handler failed:', error);
                        this.handleError(error, messageContext);
                    }
                }
            }
        });
        this.app.error(async (error) => {
            logError('Slack app error:', error);
            this.handleError(error, null);
        });
    }
    onTextMessage(handler) {
        this.messageHandlers.push(handler);
    }
    onError(handler) {
        this.errorHandlers.push(handler);
    }
    handleError(error, messageContext) {
        for (const handler of this.errorHandlers) {
            try {
                handler(error, messageContext);
            }
            catch (handlerError) {
                logError('Error handler itself failed:', handlerError);
            }
        }
    }
    async launch() {
        await this.app.start();
        logInfo('⚡️ Slack app is running!');
    }
    async sendTextToChat(chatId, text) {
        const channel = String(chatId);
        const formattedText = toSlackMrkdwn(text);
        const chunks = splitTextIntoChunks(formattedText, this.maxMessageLength);
        for (const chunk of chunks) {
            await this.app.client.chat.postMessage({
                channel,
                text: chunk,
            });
        }
    }
    stop(reason) {
        logInfo(`Stopping Slack client: ${reason}`);
        this.app.stop();
    }
}
