// clawless/dist/utils/callbackState.js
import fs from 'node:fs';
import { getErrorMessage } from './error.js';
export function ensureClawlessHomeDirectory(clawlessHomePath) {
    fs.mkdirSync(clawlessHomePath, { recursive: true });
}
export function resolveChatId(value) {
    if (value === undefined || value === null) {
        return null;
    }
    const normalized = String(value).trim();
    if (!normalized) {
        return null;
    }
    if (/^-?\d+$/.test(normalized)) {
        return normalized;
    }
    return normalized;
}
export function loadPersistedCallbackChatId(callbackStateFilePath, logInfo) {
    try {
        if (!fs.existsSync(callbackStateFilePath)) {
            return null;
        }
        const parsed = JSON.parse(fs.readFileSync(callbackStateFilePath, 'utf8'));
        return resolveChatId(parsed?.chatId);
    }
    catch (error) {
        logInfo('Failed to load callback chat state', {
            callbackChatStateFilePath: callbackStateFilePath,
            error: getErrorMessage(error),
        });
        return null;
    }
}
export function persistCallbackChatId(callbackStateFilePath, chatId, ensureHomeDirectory, logInfo) {
    try {
        ensureHomeDirectory();
        fs.writeFileSync(callbackStateFilePath, `${JSON.stringify({ chatId: String(chatId), updatedAt: new Date().toISOString() }, null, 2)}\n`, 'utf8');
    }
    catch (error) {
        logInfo('Failed to persist callback chat state', {
            callbackChatStateFilePath: callbackStateFilePath,
            error: getErrorMessage(error),
        });
    }
}
