// clawless/dist/utils/geminiMcpHelpers.js
/**
 * Helper functions to read MCP server configuration from Gemini CLI settings.
 */
import { readFileSync, existsSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';
import { logError } from './error.js';
const GEMINI_SETTINGS_PATH = join(homedir(), '.gemini', 'settings.json');
/**
 * Read MCP server names from Gemini CLI settings file.
 * This allows ACP mode to use the same MCP servers configured in normal Gemini CLI.
 */
export function getGeminiMcpServerNames() {
    try {
        if (!existsSync(GEMINI_SETTINGS_PATH)) {
            return [];
        }
        const content = readFileSync(GEMINI_SETTINGS_PATH, 'utf-8');
        const settings = JSON.parse(content);
        if (settings.mcpServers && typeof settings.mcpServers === 'object') {
            const names = Object.keys(settings.mcpServers);
            return names;
        }
        return [];
    }
    catch (error) {
        logError('Failed to read Gemini MCP server names:', error);
        return [];
    }
}
/**
 * Read full MCP server configuration from Gemini CLI settings file.
 * Returns array format compatible with ACP mcpServers parameter.
 */
export function getGeminiMcpServersForAcp() {
    try {
        if (!existsSync(GEMINI_SETTINGS_PATH)) {
            return [];
        }
        const content = readFileSync(GEMINI_SETTINGS_PATH, 'utf-8');
        const settings = JSON.parse(content);
        if (!settings.mcpServers || typeof settings.mcpServers !== 'object') {
            return [];
        }
        // Convert to ACP-compatible format
        return Object.entries(settings.mcpServers)
            .map(([name, config]) => {
            const server = config;
            if (server.command) {
                // STDIO type server
                return {
                    name,
                    command: server.command,
                    args: server.args || [],
                    env: server.env ? Object.entries(server.env).map(([key, value]) => ({ name: key, value })) : [],
                };
            }
            if (server.url) {
                // HTTP/SSE type server
                return {
                    name,
                    type: server.type || 'sse',
                    url: server.url,
                    headers: server.headers || [],
                };
            }
            return null;
        })
            .filter(Boolean);
    }
    catch (error) {
        logError('Failed to read Gemini MCP servers for ACP:', error);
        return [];
    }
}
