// clawless/dist/acp/mcpServerHelpers.js
function normalizeEnvArray(envValue) {
    if (Array.isArray(envValue)) {
        return envValue
            .filter((entry) => entry && typeof entry === 'object')
            .map((entry) => {
            const candidate = entry;
            return {
                name: String(candidate.name ?? ''),
                value: String(candidate.value ?? ''),
            };
        })
            .filter((entry) => entry.name.length > 0);
    }
    if (envValue && typeof envValue === 'object') {
        return Object.entries(envValue).map(([name, value]) => ({
            name,
            value: String(value ?? ''),
        }));
    }
    return [];
}
function normalizeSingleMcpServer(name, serverConfig) {
    if (!serverConfig || typeof serverConfig !== 'object' || Array.isArray(serverConfig)) {
        return null;
    }
    const candidate = serverConfig;
    const hasCommand = typeof candidate.command === 'string' && candidate.command.length > 0;
    const hasUrl = typeof candidate.url === 'string' && candidate.url.length > 0;
    if (hasCommand) {
        return {
            name,
            command: String(candidate.command),
            args: Array.isArray(candidate.args) ? candidate.args.map((arg) => String(arg)) : [],
            env: normalizeEnvArray(candidate.env),
        };
    }
    if (hasUrl) {
        const type = candidate.type === 'sse' ? 'sse' : 'http';
        const headers = Array.isArray(candidate.headers)
            ? candidate.headers
                .filter((header) => header && typeof header === 'object')
                .map((header) => {
                const typedHeader = header;
                return {
                    name: String(typedHeader.name ?? ''),
                    value: String(typedHeader.value ?? ''),
                };
            })
                .filter((header) => header.name.length > 0)
            : [];
        return {
            type,
            name,
            url: String(candidate.url),
            headers,
        };
    }
    return null;
}
function normalizeMcpServers(value) {
    if (Array.isArray(value)) {
        return value
            .map((entry, index) => normalizeSingleMcpServer(`server_${index + 1}`, entry))
            .filter((entry) => Boolean(entry));
    }
    if (value && typeof value === 'object') {
        return Object.entries(value)
            .map(([name, serverConfig]) => normalizeSingleMcpServer(name, serverConfig))
            .filter((entry) => Boolean(entry));
    }
    return [];
}
export function getMcpServersForSession(options) {
    const { acpMcpServersJson: raw, logInfo, getErrorMessage, invalidEnvMessage, logDetails } = options;
    if (!raw) {
        return {
            source: 'default-empty',
            mcpServers: [],
        };
    }
    try {
        return {
            source: 'env-override',
            mcpServers: normalizeMcpServers(JSON.parse(raw)),
        };
    }
    catch (error) {
        logInfo(invalidEnvMessage, {
            error: getErrorMessage(error),
            ...logDetails,
        });
        return {
            source: 'default-empty',
            mcpServers: [],
        };
    }
}
