// clawless/dist/app/CallbackServerManager.js
import { createCallbackServer } from '../core/callbackServer.js';
import { logInfo } from '../utils/error.js';
export class CallbackServerManager {
    startCallbackServer;
    stopCallbackServer;
    constructor(options) {
        const { startCallbackServer, stopCallbackServer } = createCallbackServer({
            callbackHost: options.config.CALLBACK_HOST,
            callbackPort: options.config.CALLBACK_PORT,
            callbackAuthToken: options.config.CALLBACK_AUTH_TOKEN,
            callbackMaxBodyBytes: options.config.CALLBACK_MAX_BODY_BYTES,
            cronScheduler: options.cronScheduler,
            messagingClient: options.messagingClient,
            messagingPlatform: options.config.MESSAGING_PLATFORM,
            getLastIncomingChatId: options.getLastIncomingChatId,
            semanticConversationMemory: options.semanticConversationMemory,
            conversationHistoryMaxTotalChars: options.config.CONVERSATION_HISTORY_MAX_TOTAL_CHARS,
            conversationHistoryRecapTopK: options.config.CONVERSATION_HISTORY_RECAP_TOP_K,
            logInfo,
        });
        this.startCallbackServer = startCallbackServer;
        this.stopCallbackServer = stopCallbackServer;
    }
    start() {
        this.startCallbackServer();
    }
    stop() {
        this.stopCallbackServer();
    }
}
