// clawless/dist/app/SchedulerManager.js
import { CronScheduler } from '../scheduler/cronScheduler.js';
import { createScheduledJobHandler } from '../scheduler/scheduledJobHandler.js';
import { logInfo } from '../utils/error.js';
import { normalizeOutgoingText } from '../utils/commandText.js';
export class SchedulerManager {
    cronScheduler;
    constructor(options) {
        const handleScheduledJob = createScheduledJobHandler({
            logInfo,
            buildPromptWithMemory: options.buildPromptWithMemory,
            runScheduledPromptWithTempAcp: options.runScheduledPromptWithTempAcp,
            resolveTargetChatId: options.resolveTargetChatId,
            sendTextToChat: (chatId, text) => {
                const client = options.getMessagingClient();
                if (client) {
                    return client.sendTextToChat(chatId, text);
                }
                logInfo('Warning: messagingClient not yet initialized when sending scheduled job text');
                return Promise.resolve();
            },
            normalizeOutgoingText,
            appendContextToAgent: options.appendContextToAgent,
        });
        this.cronScheduler = new CronScheduler(handleScheduledJob, {
            persistenceFilePath: options.config.SCHEDULES_FILE_PATH,
            timezone: options.config.TZ,
            logInfo,
        });
        logInfo('Scheduler persistence configured', {
            schedulesFilePath: options.config.SCHEDULES_FILE_PATH,
        });
    }
    getCronScheduler() {
        return this.cronScheduler;
    }
    shutdown() {
        this.cronScheduler.shutdown();
    }
}
